src/Tasks/Exec.cs

```diff
@@ -609,20 +612,53 @@ protected internal override void AddCommandLineCommands(CommandLineBuilderExtens
                     }
                     commandLine.AppendSwitch("/C"); // run then terminate
 
-                    // If for some crazy reason the path has a & character and a space in it
-                    // then get the short path of the temp path, which should not have spaces in it
-                    // and then escape the &
-                    if (batchFileForCommandLine.Contains("&") && !batchFileForCommandLine.Contains("^&"))
+                    if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
                     {
-                        batchFileForCommandLine = NativeMethodsShared.GetShortFilePath(batchFileForCommandLine);
-                        batchFileForCommandLine = batchFileForCommandLine.Replace("&", "^&");
+                        StringBuilder fileName = StringBuilderCache.Acquire(batchFileForCommandLine.Length);
+
+                        // Escape special characters that need to be escaped.
+                        for (int i = 0; i < batchFileForCommandLine.Length; i++)
+                        {
+                            char c = batchFileForCommandLine[i];
+
+                            if (ShouldEscapeCharacter(c) && (i == 0 || batchFileForCommandLine[i - 1] != '^'))
+                            {
+                                fileName.Append('^');
```

Comment:
nit: If you wanted to squeeze a bit more perf out of this, you could acquire the StringBuilder first time you hit this line. Then in the common case of no escaping, there would be no StringBuilder and you wouldn't allocate a new string (allocation is happening inside `StringBuilderCache.GetStringAndRelease`).